File: zh-hant/codes/dart/chapter_dynamic_programming/edit_distance.dart
/**
 * File: edit_distance.dart
 * Created Time: 2023-08-11
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

import 'dart:math';

/* 編輯距離：暴力搜尋 */
int editDistanceDFS(String s, String t, int i, int j) {
  // 若 s 和 t 都為空，則返回 0
  if (i == 0 && j == 0) return 0;
  // 若 s 為空，則返回 t 長度
  if (i == 0) return j;
  // 若 t 為空，則返回 s 長度
  if (j == 0) return i;
  // 若兩字元相等，則直接跳過此兩字元
  if (s[i - 1] == t[j - 1]) return editDistanceDFS(s, t, i - 1, j - 1);
  // 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
  int insert = editDistanceDFS(s, t, i, j - 1);
  int delete = editDistanceDFS(s, t, i - 1, j);
  int replace = editDistanceDFS(s, t, i - 1, j - 1);
  // 返回最少編輯步數
  return min(min(insert, delete), replace) + 1;
}

/* 編輯距離：記憶化搜尋 */
int editDistanceDFSMem(String s, String t, List<List<int>> mem, int i, int j) {
  // 若 s 和 t 都為空，則返回 0
  if (i == 0 && j == 0) return 0;
  // 若 s 為空，則返回 t 長度
  if (i == 0) return j;
  // 若 t 為空，則返回 s 長度
  if (j == 0) return i;
  // 若已有記錄，則直接返回之
  if (mem[i][j] != -1) return mem[i][j];
  // 若兩字元相等，則直接跳過此兩字元
  if (s[i - 1] == t[j - 1]) return editDistanceDFSMem(s, t, mem, i - 1, j - 1);
  // 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
  int insert = editDistanceDFSMem(s, t, mem, i, j - 1);
  int delete = editDistanceDFSMem(s, t, mem, i - 1, j);
  int replace = editDistanceDFSMem(s, t, mem, i - 1, j - 1);
  // 記錄並返回最少編輯步數
  mem[i][j] = min(min(insert, delete), replace) + 1;
  return mem[i][j];
}

/* 編輯距離：動態規劃 */
int editDistanceDP(String s, String t) {
  int n = s.length, m = t.length;
  List<List<int>> dp = List.generate(n + 1, (_) => List.filled(m + 1, 0));
  // 狀態轉移：首行首列
  for (int i = 1; i <= n; i++) {
    dp[i][0] = i;
  }
  for (int j = 1; j <= m; j++) {
    dp[0][j] = j;
  }
  // 狀態轉移：其餘行和列
  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= m; j++) {
      if (s[i - 1] == t[j - 1]) {
        // 若兩字元相等，則直接跳過此兩字元
        dp[i][j] = dp[i - 1][j - 1];
      } else {
        // 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
        dp[i][j] = min(min(dp[i][j - 1], dp[i - 1][j]), dp[i - 1][j - 1]) + 1;
      }
    }
  }
  return dp[n][m];
}

/* 編輯距離：空間最佳化後的動態規劃 */
int editDistanceDPComp(String s, String t) {
  int n = s.length, m = t.length;
  List<int> dp = List.filled(m + 1, 0);
  // 狀態轉移：首行
  for (int j = 1; j <= m; j++) {
    dp[j] = j;
  }
  // 狀態轉移：其餘行
  for (int i = 1; i <= n; i++) {
    // 狀態轉移：首列
    int leftup = dp[0]; // 暫存 dp[i-1, j-1]
    dp[0] = i;
    // 狀態轉移：其餘列
    for (int j = 1; j <= m; j++) {
      int temp = dp[j];
      if (s[i - 1] == t[j - 1]) {
        // 若兩字元相等，則直接跳過此兩字元
        dp[j] = leftup;
      } else {
        // 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
        dp[j] = min(min(dp[j - 1], dp[j]), leftup) + 1;
      }
      leftup = temp; // 更新為下一輪的 dp[i-1, j-1]
    }
  }
  return dp[m];
}

/* Driver Code */
void main() {
  String s = "bag";
  String t = "pack";
  int n = s.length, m = t.length;

  // 暴力搜尋
  int res = editDistanceDFS(s, t, n, m);
  print("將 " + s + " 更改為 " + t + " 最少需要編輯 $res 步");

  // 記憶化搜尋
  List<List<int>> mem = List.generate(n + 1, (_) => List.filled(m + 1, -1));
  res = editDistanceDFSMem(s, t, mem, n, m);
  print("將 " + s + " 更改為 " + t + " 最少需要編輯 $res 步");

  // 動態規劃
  res = editDistanceDP(s, t);
  print("將 " + s + " 更改為 " + t + " 最少需要編輯 $res 步");

  // 空間最佳化後的動態規劃
  res = editDistanceDPComp(s, t);
  print("將 " + s + " 更改為 " + t + " 最少需要編輯 $res 步");
}

File: zh-hant/codes/ruby/chapter_dynamic_programming/edit_distance.rb
=begin
File: edit_distance.rb
Created Time: 2024-05-29
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### 編輯距離：暴力搜尋 ###
def edit_distance_dfs(s, t, i, j)
  # 若 s 和 t 都為空，則返回 0
  return 0 if i == 0 && j == 0
  # 若 s 為空，則返回 t 長度
  return j if i == 0
  # 若 t 為空，則返回 s 長度
  return i if j == 0
  # 若兩字元相等，則直接跳過此兩字元
  return edit_distance_dfs(s, t, i - 1, j - 1) if s[i - 1] == t[j - 1]
  # 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
  insert = edit_distance_dfs(s, t, i, j - 1)
  delete = edit_distance_dfs(s, t, i - 1, j)
  replace = edit_distance_dfs(s, t, i - 1, j - 1)
  # 返回最少編輯步數
  [insert, delete, replace].min + 1
end

def edit_distance_dfs_mem(s, t, mem, i, j)
  # 若 s 和 t 都為空，則返回 0
  return 0 if i == 0 && j == 0
  # 若 s 為空，則返回 t 長度
  return j if i == 0
  # 若 t 為空，則返回 s 長度
  return i if j == 0
  # 若已有記錄，則直接返回之
  return mem[i][j] if mem[i][j] != -1
  # 若兩字元相等，則直接跳過此兩字元
  return edit_distance_dfs_mem(s, t, mem, i - 1, j - 1) if s[i - 1] == t[j - 1]
  # 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
  insert = edit_distance_dfs_mem(s, t, mem, i, j - 1)
  delete = edit_distance_dfs_mem(s, t, mem, i - 1, j)
  replace = edit_distance_dfs_mem(s, t, mem, i - 1, j - 1)
  # 記錄並返回最少編輯步數
  mem[i][j] = [insert, delete, replace].min + 1
end

### 編輯距離：動態規劃 ###
def edit_distance_dp(s, t)
  n, m = s.length, t.length
  dp = Array.new(n + 1) { Array.new(m + 1, 0) }
  # 狀態轉移：首行首列
  (1...(n + 1)).each { |i| dp[i][0] = i }
  (1...(m + 1)).each { |j| dp[0][j] = j }
  # 狀態轉移：其餘行和列
  for i in 1...(n + 1)
    for j in 1...(m +1)
      if s[i - 1] == t[j - 1]
        # 若兩字元相等，則直接跳過此兩字元
        dp[i][j] = dp[i - 1][j - 1]
      else
        # 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
        dp[i][j] = [dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]].min + 1
      end
    end
  end
  dp[n][m]
end

### 編輯距離：空間最佳化後的動態規劃 ###
def edit_distance_dp_comp(s, t)
  n, m = s.length, t.length
  dp = Array.new(m + 1, 0)
  # 狀態轉移：首行
  (1...(m + 1)).each { |j| dp[j] = j }
  # 狀態轉移：其餘行
  for i in 1...(n + 1)
    # 狀態轉移：首列
    leftup = dp.first # 暫存 dp[i-1, j-1]
    dp[0] += 1
    # 狀態轉移：其餘列
    for j in 1...(m + 1)
      temp = dp[j]
      if s[i - 1] == t[j - 1]
        # 若兩字元相等，則直接跳過此兩字元
        dp[j] = leftup
      else
        # 最少編輯步數 = 插入、刪除、替換這三種操作的最少編輯步數 + 1
        dp[j] = [dp[j - 1], dp[j], leftup].min + 1
      end
      leftup = temp # 更新為下一輪的 dp[i-1, j-1]
    end
  end
  dp[m]
end

### Driver Code ###
if __FILE__ == $0
  s = 'bag'
  t = 'pack'
  n, m = s.length, t.length

  # 暴力搜尋
  res = edit_distance_dfs(s, t, n, m)
  puts "將 #{s} 更改為 #{t} 最少需要編輯 #{res} 步"

  # 記憶化搜尋
  mem = Array.new(n + 1) { Array.new(m + 1, -1) }
  res = edit_distance_dfs_mem(s, t, mem, n, m)
  puts "將 #{s} 更改為 #{t} 最少需要編輯 #{res} 步"

  # 動態規劃
  res = edit_distance_dp(s, t)
  puts "將 #{s} 更改為 #{t} 最少需要編輯 #{res} 步"

  # 空間最佳化後的動態規劃
  res = edit_distance_dp_comp(s, t)
  puts "將 #{s} 更改為 #{t} 最少需要編輯 #{res} 步"
end
